LmDataset.py

```diff
@@ -1038,72 +1038,116 @@ class TranslationDataset(CachedDataset2):
       target.vocab.pkl
   """
 
-  MapToDataKeys = {"source": "data", "target": "classes"}  # just by our convention
-  _main_data_key = None
-  _main_classes_key = None
+  source_file_prefix = "source"
+  target_file_prefix = "target"
+  main_source_data_key = "data"
+  main_target_data_key = "classes"
 
   def __init__(self, path, file_postfix, source_postfix="", target_postfix="",
                source_only=False,
                unknown_label=None,
+               source_factors=None,
+               target_factors=None,
+               factor_separator='|',
                seq_list_file=None,
                use_cache_manager=False,
                **kwargs):
     """
     :param str path: the directory containing the files
     :param str file_postfix: e.g. "train" or "dev". it will then search for "source." + postfix and "target." + postfix.
     :param bool random_shuffle_epoch1: if True, will also randomly shuffle epoch 1. see self.init_seq_order().
-    :param None|str source_postfix: will concat this at the end of the source. e.g.
-    :param None|str target_postfix: will concat this at the end of the target.
+    :param None|str source_postfix: will append this token to the source words (for all source factors).
+    :param None|str target_postfix: will append this token to the target words (for all target factors).
       You might want to add some sentence-end symbol.
     :param bool source_only: if targets are not available
-    :param str|None unknown_label: "UNK" or so. if not given, then will not replace unknowns but throw an error
+    :param str|dict[str,str]|None unknown_label: Label to replace out-of-vocabulary words with, e.g. "<UNK>".
+      If not given, will not replace unknowns but throw an error. Can also be a dict data_key -> unknown_label
+      to configure for each data key separately (default for each key is None).
+    :param list[str]|None source_factors: Names of source factors (excluding first factor, which is always
+      called 'data'). Words in source file have to have that many factors. Also, a vocabulary
+      "<factor_name>.voc.pkl" has to exist for each factor.
```

Comment:
What does that mean? These are becoming data stream then?
These data streams are always the same length as `data` (or `classes` respectively)? This is important information, which should be documented here. Also, you should give some actual examples, and maybe ref to papers, and other software (you said other software uses the same format).